File: Kernel Simulator/Network/FTP/FTPShellCommon.vb

'    Kernel Simulator  Copyright (C) 2018-2022  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.Threading

'TODO: Unify GetLine as we're now moving to IShell to handle shells
Public Module FTPShellCommon

    Public ReadOnly FTPCommands As New Dictionary(Of String, CommandInfo) From {{"connect", New CommandInfo("connect", ShellType.FTPShell, "Connects to an FTP server (it must start with ""ftp://"" or ""ftps://"")", {"<server>"}, True, 1, New FTP_ConnectCommand)},
                                                                                {"cdl", New CommandInfo("cdl", ShellType.FTPShell, "Changes local directory to download to or upload from", {"<directory>"}, True, 1, New FTP_CdlCommand)},
                                                                                {"cdr", New CommandInfo("cdr", ShellType.FTPShell, "Changes remote directory to download from or upload to", {"<directory>"}, True, 1, New FTP_CdrCommand)},
                                                                                {"cp", New CommandInfo("cp", ShellType.FTPShell, "Copies file or directory to another file or directory.", {"<sourcefileordir> <targetfileordir>"}, True, 2, New FTP_CpCommand)},
                                                                                {"del", New CommandInfo("del", ShellType.FTPShell, "Deletes remote file from server", {"<file>"}, True, 1, New FTP_DelCommand)},
                                                                                {"disconnect", New CommandInfo("disconnect", ShellType.FTPShell, "Disconnects from server", {"[-f]"}, False, 0, New FTP_DisconnectCommand, False, False, False, False, False, New Action(AddressOf (New FTP_DisconnectCommand).HelpHelper))},
                                                                                {"execute", New CommandInfo("execute", ShellType.FTPShell, "Executes an FTP server command", {"<command>"}, True, 1, New FTP_ExecuteCommand)},
                                                                                {"exit", New CommandInfo("exit", ShellType.FTPShell, "Exits FTP shell and returns to kernel", {}, False, 0, New FTP_ExitCommand)},
                                                                                {"get", New CommandInfo("get", ShellType.FTPShell, "Downloads remote file to local directory using binary or text", {"<file> [output]"}, True, 1, New ZipShell_GetCommand)},
                                                                                {"getfolder", New CommandInfo("getfolder", ShellType.FTPShell, "Downloads remote folder to local directory using binary or text", {"<folder> [outputfolder]"}, True, 1, New FTP_GetFolderCommand)},
                                                                                {"help", New CommandInfo("help", ShellType.FTPShell, "Shows help screen", {"[command]"}, False, 0, New FTP_HelpCommand)},
                                                                                {"info", New CommandInfo("info", ShellType.FTPShell, "FTP server information", {}, False, 0, New FTP_InfoCommand)},
                                                                                {"lsl", New CommandInfo("lsl", ShellType.FTPShell, "Lists local directory", {"[-showdetails|-suppressmessages] [dir]"}, False, 0, New FTP_LslCommand, False, False, False, False, False, New Action(AddressOf (New FTP_LslCommand).HelpHelper))},
                                                                                {"lsr", New CommandInfo("lsr", ShellType.FTPShell, "Lists remote directory", {"[-showdetails] [dir]"}, False, 0, New FTP_LsrCommand, False, False, False, False, False, New Action(AddressOf (New FTP_LsrCommand).HelpHelper))},
                                                                                {"mv", New CommandInfo("mv", ShellType.FTPShell, "Moves file or directory to another file or directory. You can also use that to rename files.", {"<sourcefileordir> <targetfileordir>"}, True, 2, New FTP_MvCommand)},
                                                                                {"put", New CommandInfo("put", ShellType.FTPShell, "Uploads local file to remote directory using binary or text", {"<file> [output]"}, True, 1, New FTP_PutCommand)},
                                                                                {"putfolder", New CommandInfo("putfolder", ShellType.FTPShell, "Uploads local folder to remote directory using binary or text", {"<folder> [outputfolder]"}, True, 1, New FTP_PutFolderCommand)},
                                                                                {"pwdl", New CommandInfo("pwdl", ShellType.FTPShell, "Gets current local directory", {}, False, 0, New FTP_PwdlCommand)},
                                                                                {"pwdr", New CommandInfo("pwdr", ShellType.FTPShell, "Gets current remote directory", {}, False, 0, New FTP_PwdrCommand)},
                                                                                {"perm", New CommandInfo("perm", ShellType.FTPShell, "Sets file permissions. This is supported only on FTP servers that run Unix.", {"<file> <permnumber>"}, True, 2, New FTP_PermCommand)},
                                                                                {"quickconnect", New CommandInfo("quickconnect", ShellType.FTPShell, "Uses information from Speed Dial to connect to any network quickly", {}, False, 0, New FTP_QuickConnectCommand)},
                                                                                {"sumfile", New CommandInfo("sumfile", ShellType.FTPShell, "Calculates file sums.", {"<file> <MD5/SHA1/SHA256/SHA512/CRC>"}, True, 2, New FTP_SumFileCommand)},
                                                                                {"sumfiles", New CommandInfo("sumfiles", ShellType.FTPShell, "Calculates sums of files in specified directory.", {"<file> <MD5/SHA1/SHA256/SHA512/CRC>"}, True, 2, New FTP_SumFilesCommand)},
                                                                                {"type", New CommandInfo("type", ShellType.FTPShell, "Sets the type for this session", {"<a/b>"}, True, 1, New FTP_TypeCommand)}}
    Public FtpConnected As Boolean
    Public FtpSite As String
    Public FtpCurrentDirectory As String
    Public FtpCurrentRemoteDir As String
    Public FtpUser As String
    Public FTPModCommands As New ArrayList
    Public FTPShellPromptStyle As String = ""
    Public ClientFTP As FtpClient
    Public FtpShowDetailsInList As Boolean = True
    Public FtpUserPromptStyle As String = ""
    Public FtpPassPromptStyle As String = ""
    Public FtpUseFirstProfile As Boolean
    Public FtpNewConnectionsToSpeedDial As Boolean = True
    Public FtpTryToValidateCertificate As Boolean = True
    Public FtpRecursiveHashing As Boolean
    Public FtpShowMotd As Boolean = True
    Public FtpAlwaysAcceptInvalidCerts As Boolean
    Public FtpVerifyRetryAttempts As Integer = 3
    Public FtpConnectTimeout As Integer = 15000
    Public FtpDataConnectTimeout As Integer = 15000
    Public FtpProtocolVersions As FtpIpVersion = FtpIpVersion.ANY
    Friend FtpPass As String

    ''' <summary>
    ''' Parses a command line from FTP shell
    ''' </summary>
    Public Sub FTPGetLine(FtpCommand As String)
        Dim words As String() = FtpCommand.SplitEncloseDoubleQuotes(" ")
        Wdbg(DebugLevel.I, $"Is the command found? {FTPCommands.ContainsKey(words(0))}")
        If FTPCommands.ContainsKey(words(0)) Then
            Wdbg(DebugLevel.I, "Command found.")
            Dim Params As New ExecuteCommandThreadParameters(FtpCommand, ShellType.FTPShell, Nothing)
            FTPStartCommandThread = New Thread(AddressOf ExecuteCommand) With {.Name = "FTP Command Thread"}
            FTPStartCommandThread.Start(Params)
            FTPStartCommandThread.Join()
        ElseIf FTPModCommands.Contains(words(0)) Then
            Wdbg(DebugLevel.I, "Mod command found.")
            ExecuteModCommand(FtpCommand)
        ElseIf FTPShellAliases.Keys.Contains(words(0)) Then
            Wdbg(DebugLevel.I, "FTP shell alias command found.")
            FtpCommand = FtpCommand.Replace($"""{words(0)}""", words(0))
            ExecuteFTPAlias(FtpCommand)
        ElseIf Not FtpCommand.StartsWith(" ") Then
            Wdbg(DebugLevel.E, "Command {0} not found.", FtpCommand)
            Write(DoTranslation("FTP message: The requested command {0} is not found. See 'help' for a list of available commands specified on FTP shell."), True, ColTypes.Error, words(0))
        End If
    End Sub

End Module

File: Kernel Simulator/Shell/Shells/FTP/FTPShellCommon.cs
﻿
// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System.Collections.Generic;
using FluentFTP;
using KS.Shell.ShellBase.Commands;
using KS.Shell.ShellBase.Shells;
using KS.Shell.Shells.FTP.Commands;

namespace KS.Shell.Shells.FTP
{
    /// <summary>
    /// Common FTP shell class
    /// </summary>
    public static class FTPShellCommon
    {

        /// <summary>
        /// FTP commands
        /// </summary>
        public readonly static Dictionary<string, CommandInfo> FTPCommands = new()
        {
            { "connect", new CommandInfo("connect", ShellType.FTPShell, "Connects to an FTP server (it must start with \"ftp://\" or \"ftps://\")", new CommandArgumentInfo(new[] { "<server>" }, true, 1), new FTP_ConnectCommand()) },
            { "cdl", new CommandInfo("cdl", ShellType.FTPShell, "Changes local directory to download to or upload from", new CommandArgumentInfo(new[] { "<directory>" }, true, 1), new FTP_CdlCommand()) },
            { "cdr", new CommandInfo("cdr", ShellType.FTPShell, "Changes remote directory to download from or upload to", new CommandArgumentInfo(new[] { "<directory>" }, true, 1), new FTP_CdrCommand()) },
            { "cp", new CommandInfo("cp", ShellType.FTPShell, "Copies file or directory to another file or directory.", new CommandArgumentInfo(new[] { "<sourcefileordir> <targetfileordir>" }, true, 2), new FTP_CpCommand()) },
            { "del", new CommandInfo("del", ShellType.FTPShell, "Deletes remote file from server", new CommandArgumentInfo(new[] { "<file>" }, true, 1), new FTP_DelCommand()) },
            { "disconnect", new CommandInfo("disconnect", ShellType.FTPShell, "Disconnects from server", new CommandArgumentInfo(new[] { "[-f]" }, false, 0), new FTP_DisconnectCommand()) },
            { "execute", new CommandInfo("execute", ShellType.FTPShell, "Executes an FTP server command", new CommandArgumentInfo(new[] { "<command>" }, true, 1), new FTP_ExecuteCommand()) },
            { "get", new CommandInfo("get", ShellType.FTPShell, "Downloads remote file to local directory using binary or text", new CommandArgumentInfo(new[] { "<file> [output]" }, true, 1), new FTP_GetCommand()) },
            { "getfolder", new CommandInfo("getfolder", ShellType.FTPShell, "Downloads remote folder to local directory using binary or text", new CommandArgumentInfo(new[] { "<folder> [outputfolder]" }, true, 1), new FTP_GetFolderCommand()) },
            { "info", new CommandInfo("info", ShellType.FTPShell, "FTP server information", new CommandArgumentInfo(), new FTP_InfoCommand()) },
            { "lsl", new CommandInfo("lsl", ShellType.FTPShell, "Lists local directory", new CommandArgumentInfo(new[] { "[-showdetails|-suppressmessages] [dir]" }, false, 0), new FTP_LslCommand()) },
            { "lsr", new CommandInfo("lsr", ShellType.FTPShell, "Lists remote directory", new CommandArgumentInfo(new[] { "[-showdetails] [dir]" }, false, 0), new FTP_LsrCommand()) },
            { "mv", new CommandInfo("mv", ShellType.FTPShell, "Moves file or directory to another file or directory. You can also use that to rename files.", new CommandArgumentInfo(new[] { "<sourcefileordir> <targetfileordir>" }, true, 2), new FTP_MvCommand()) },
            { "put", new CommandInfo("put", ShellType.FTPShell, "Uploads local file to remote directory using binary or text", new CommandArgumentInfo(new[] { "<file> [output]" }, true, 1), new FTP_PutCommand()) },
            { "putfolder", new CommandInfo("putfolder", ShellType.FTPShell, "Uploads local folder to remote directory using binary or text", new CommandArgumentInfo(new[] { "<folder> [outputfolder]" }, true, 1), new FTP_PutFolderCommand()) },
            { "pwdl", new CommandInfo("pwdl", ShellType.FTPShell, "Gets current local directory", new CommandArgumentInfo(), new FTP_PwdlCommand()) },
            { "pwdr", new CommandInfo("pwdr", ShellType.FTPShell, "Gets current remote directory", new CommandArgumentInfo(), new FTP_PwdrCommand()) },
            { "perm", new CommandInfo("perm", ShellType.FTPShell, "Sets file permissions. This is supported only on FTP servers that run Unix.", new CommandArgumentInfo(new[] { "<file> <permnumber>" }, true, 2), new FTP_PermCommand()) },
            { "quickconnect", new CommandInfo("quickconnect", ShellType.FTPShell, "Uses information from Speed Dial to connect to any network quickly", new CommandArgumentInfo(), new FTP_QuickConnectCommand()) },
            { "sumfile", new CommandInfo("sumfile", ShellType.FTPShell, "Calculates file sums.", new CommandArgumentInfo(new[] { "<file> <MD5/SHA1/SHA256/SHA512/CRC>" }, true, 2), new FTP_SumFileCommand()) },
            { "sumfiles", new CommandInfo("sumfiles", ShellType.FTPShell, "Calculates sums of files in specified directory.", new CommandArgumentInfo(new[] { "<file> <MD5/SHA1/SHA256/SHA512/CRC>" }, true, 2), new FTP_SumFilesCommand()) },
            { "type", new CommandInfo("type", ShellType.FTPShell, "Sets the type for this session", new CommandArgumentInfo(new[] { "<a/b>" }, true, 1), new FTP_TypeCommand()) }
        };
        /// <summary>
        /// FTP current local directory
        /// </summary>
        public static string FtpCurrentDirectory;
        /// <summary>
        /// FTP current remote directory
        /// </summary>
        public static string FtpCurrentRemoteDir;
        /// <summary>
        /// FTP show details in list
        /// </summary>
        public static bool FtpShowDetailsInList = true;
        /// <summary>
        /// FTP user prompt style
        /// </summary>
        public static string FtpUserPromptStyle = "";
        /// <summary>
        /// FTP password prompt style
        /// </summary>
        public static string FtpPassPromptStyle = "";
        /// <summary>
        /// FTP always use first profile
        /// </summary>
        public static bool FtpUseFirstProfile;
        /// <summary>
        /// FTP add new connections to speed dial
        /// </summary>
        public static bool FtpNewConnectionsToSpeedDial = true;
        /// <summary>
        /// FTP try to validate the certificate
        /// </summary>
        public static bool FtpTryToValidateCertificate = true;
        /// <summary>
        /// FTP recursive hashing
        /// </summary>
        public static bool FtpRecursiveHashing;
        /// <summary>
        /// FTP show MOTD
        /// </summary>
        public static bool FtpShowMotd = true;
        /// <summary>
        /// FTP always accept invalid certificates. Turning it on is not recommended.
        /// </summary>
        public static bool FtpAlwaysAcceptInvalidCerts;
        /// <summary>
        /// FTP verify retry attempts
        /// </summary>
        public static int FtpVerifyRetryAttempts = 3;
        /// <summary>
        /// FTP connection timeout in milliseconds
        /// </summary>
        public static int FtpConnectTimeout = 15000;
        /// <summary>
        /// FTP data connection timeout in milliseconds
        /// </summary>
        public static int FtpDataConnectTimeout = 15000;
        /// <summary>
        /// FTP protocol versions
        /// </summary>
        public static FtpIpVersion FtpProtocolVersions = FtpIpVersion.ANY;
        internal static FtpClient _clientFTP;
        internal static bool FtpConnected;
        internal static string FtpSite;
        internal static string FtpPass;
        internal static string FtpUser;
        internal readonly static Dictionary<string, CommandInfo> FTPModCommands = new();

        /// <summary>
        /// The FTP client used to connect to the FTP server
        /// </summary>
        public static FtpClient ClientFTP
        {
            get
            {
                return _clientFTP;
            }
        }

    }
}
